stake-pool/single-pool/tests/accounts.rs

```diff
@@ -0,0 +1,177 @@
+            &context.payer,
+            &context.last_blockhash,
+        )
+        .await;
+
+        instruction::initialize(
+            &id(),
+            &accounts.vote_account.pubkey(),
+            &accounts.alice.pubkey(),
+            &rent,
+            minimum_delegation,
+        )
+    } else {
+        accounts
+            .initialize_for_deposit(context, TEST_STAKE_AMOUNT, None)
+            .await;
+        advance_epoch(context).await;
+
+        vec![]
+    };
+
+    let deposit_instructions = instruction::deposit(
+        &id(),
+        &accounts.vote_account.pubkey(),
+        &accounts.alice_stake.pubkey(),
+        &accounts.alice_token,
+        &accounts.alice.pubkey(),
+        &accounts.alice.pubkey(),
+    );
+
+    let withdraw_instructions = if test_mode == TestMode::Withdraw {
+        let message = Message::new(&deposit_instructions, Some(&accounts.alice.pubkey()));
+        let transaction = Transaction::new(&[&accounts.alice], message, context.last_blockhash);
+
+        context
+            .banks_client
+            .process_transaction(transaction)
+            .await
+            .unwrap();
+
+        create_blank_stake_account(
+            &mut context.banks_client,
+            &accounts.alice,
+            &context.last_blockhash,
+            &accounts.alice_stake,
+        )
+        .await;
+
+        instruction::withdraw(
+            &id(),
+            &accounts.vote_account.pubkey(),
+            &accounts.alice_stake.pubkey(),
+            &accounts.alice.pubkey(),
+            &accounts.alice_token,
+            &accounts.alice.pubkey(),
+            get_token_balance(&mut context.banks_client, &accounts.alice_token).await,
+        )
+    } else {
+        vec![]
+    };
+
+    let (instructions, i) = match test_mode {
+        TestMode::Initialize => (initialize_instructions, 2),
+        TestMode::Deposit => (deposit_instructions, 2),
+        TestMode::Withdraw => (withdraw_instructions, 1),
+    };
+
+    // guard against instructions moving with code changes
+    assert_eq!(instructions[i].program_id, id());
+
+    (instructions, i)
+}
+
+#[test_case(TestMode::Initialize; "initialize")]
+#[test_case(TestMode::Deposit; "deposit")]
+#[test_case(TestMode::Withdraw; "withdraw")]
+#[tokio::test]
+async fn fail_account_checks(test_mode: TestMode) {
+    let mut context = program_test().start_with_context().await;
+    let accounts = SinglePoolAccounts::default();
+    let (instructions, i) = build_instructions(&mut context, &accounts, test_mode).await;
+
+    for j in 0..instructions[i].accounts.len() {
+        let mut instructions = instructions.clone();
+        let instruction_account = &mut instructions[i].accounts[j];
+
+        // wallet address can be arbitrary
+        if instruction_account.pubkey == accounts.alice.pubkey() {
+            continue;
+        }
+
+        let prev_pubkey = instruction_account.pubkey;
+        instruction_account.pubkey = Keypair::new().pubkey();
```

Comment:
nit: Can use `Pubkey::new_unique()` here too
```suggestion
        instruction_account.pubkey = Pubkey::new_unique();
```